### ISSUE TITLE
TwinBasic to allow CPU prioritisation of specified background timer tasks.

### ISSUE THREAD

**Issue by yereverluvinunclebert** [twinbasic/twinbasic#1978]
As many of us already know, the VB6 timers run in a 'sort of' paralellism, ie. they run in the background but VB6 has some sort of time-slicing where it allocates priority to the UI code and user interaction whilst any code in timers rund in the background with reduced access to the CPU. 

So, my suggestion. The ability to increase priority of a specific timer so that it runs at the same time-slicing priority as the main UI. At the moment a user control being manipulated slows any timer code significantly. I understand that when multi-threading becomes available a lot of tasks can be farmed out to new threads but some internal prioritisation of timers could be very useful.

Just a thought. 

This facility would be very useful in some of my existing projects allowing smoother timer animations without a major recode using multi-threading.

**Comment by dmrvb** [comment]
Yes, I do clever things with timers to give the user the impression of multi-threading/parallel processing.  Something like this could be really useful. Lets get tBv1 all up and running and then we could have a list of required enhancements.  My timer stuff does not work quite right in tB at the moment, I will do a bug report when I can work out whats going on.

**Comment by yereverluvinunclebert** [comment]
What do your timers do differently?

**Comment by dmrvb** [comment]
Something is stopping the timer in tB that does not stop it in VB6, but its a biggish project with a big third party DLL so will take some tracking down. There are lots of other issues to sort out first, including some new ones in the latest releases. Will try to find a bit of time this evening.

**Comment by yereverluvinunclebert** [comment]
OK, back to timer prioritisation.

As I think about this, the more important and useful this request seems to become. For TwinBasic to be useful in animation contexts, events need to occur either in separate threads or at least with the same priority.  If any background animation stops entirely whilst the user is still interacting with a control then this rules out TB being used for any meaningful animated graphical work and reduces its use back to simple traditional forms. 

I have used other environments that have similar timer constructs but where the timers do not run with a lower time slice. Current computers with all their CPU capability have enough capacity to run both the main UI context and the timer contexts at the same priority so they each receive equal time slices without any deterioration in response. I suppose VB6, being from the single core era of 500mhz-ish CPUs found a way to deal with timers taking all the CPU and blocking the main UI. 

This 'priority' must be being determined internally, I don't know which construct is performing this but the real-time time slicing must be being carried out by something.  Perhaps all timers have a lower priority but I'm guessing that there must be a task queue of some sort. That implies that there is a priority stored somewhere and IF the priority can be modified then this property really could do with being exposed.

We could then experiment with timers giving them equal access to CPU cycles and see how the resulting increase affects the UI. By default, they would remain at a lower value (1-3) than the current UI (4) and they could  have time slicing more approrpitae to their actual needs. 

Timers that control graphical elements could even be boosted dynamically according to UI load... If we had access to the priority level we could do that ourselves.

**Comment by dmrvb** [comment]
Not sure I am correct, there are others here who know more than me, but I don't think the VB6 timer is an interrupt with any variable priority, I think it works by a message in the message queue so just gets processed in the same way as the other GUI messages.

**Comment by yereverluvinunclebert** [comment]
I think you may well be correct but timers do receive less time-slicing of CPU somehow specifically when GUI elements are being manipulated by the user. In practical terms, the timers receive a lower priority access to CPU.

I believe there is a message queue of some sort and the allocation oif tasks from that queue are managed, the end result being some sort of priority switching logic.  I'm really thinking about TwinBasic and how it might be implementing similar logic.  My suggestion is that a priority setting of 1-3 would be very useful for reducing time slicing with level 4 for user-interaction and timers running at equal priority for animation.

**Comment by yereverluvinunclebert** [comment]
Just reading this thread on the vbforum and the comments from LaVolpe explain a lot of what I am seeing and I suggest that the current timers cannot change their time slicing characteristics as they use lowest priority WM_TIMER messages.

https://www.vbforums.com/showthread.php?662466-VB6-Timers-failing-with-Windows-7

To implement a higher priority/configurable timer it would have to use a different method of messaging and that might be a compatibility-breaking issue for current timer functionality.

The only way to implement such a timer wouldmost likely be, to have a second type of non-standard timer that uses a different messaging from WM_TIMER.

**Comment by KubaO** [comment]
> the VB6 timers run in a 'sort of' parallelism, ie. they run in the background but VB6 has some sort of internal time-slicing where it allocates priority to the UI code (and user interaction) whilst any code in timers runs in the background with reduced access to the CPU.

I don't believe that this is a VB6 thing. Any system with an event loop (a.k.a. message pump) exhibits this behavior. The "run when idle" timers that have their timeout set to 0 can be used to "multitask" in a single thread. The event loop processes all other messages before it services the idle timer messages.

The code in the event handler must do a "unit of work" and cannot run continuously or the UI will hang. The big idea here is that *nothing* happens in the UI without an event (a.k.a. message) of some sort causing it. Whenever *any* user code runs, the event loop is always on the call stack. So the event loop is inherently a gatekeeper for running anything and everything. Conversely, while an event is handled, no other events get handled unless the event loop is re-entered, e.g. when a modal native dialog is shown.

It is possible to add prioritization to all event sources, since timers are not special in that respect. But that would be a serious new feature, and thus not really of priority at the moment, while on the road to version 1.0.

> What about timers based on `CreateWaitableTimer`?

Nothing special, I don't think. They still need to get dispatched from the message pump, since the pump waits not only for messages but for handles to become signaled. They are a lower level primitive, with a bit less overhead since they don't modify the state of the message queue. The timer handle itself maintains the signaled/non-signaled state, without a need to post a message to the queue.
